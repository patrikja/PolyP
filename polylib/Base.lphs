%% Literate PolyP + LaTeX script for use with lhs2TeX -*- Mode: latex; -*-
% ------------------------------------------------------------
% Basic polytypic functions
%   980409 Patrik Jansson
%   031001 Ulf Norell
%   040615 Patrik Jansson
% ------------------------------------------------------------

> module Base(pmap,fmap2,cata,ana,hylo,para,(-*-),(-+-)) where
> import PolyPrelude

%{
%align 26

> pmap :: Regular d => (a -> b) -> d a -> d b
> pmap f   = inn . fmap2 f  (pmap f)   . out
> 
> polytypic fmap2 :: (a -> c) -> (b -> d) -> f a b -> f c d
>   = \p r -> case f of
>               g + h    ->  (fmap2 p r) -+- (fmap2 p r)
>               g * h    ->  (fmap2 p r) -*- (fmap2 p r)
>               Empty    ->  \EmptyF -> EmptyF
>               Par      ->  ParF . p . unParF
>               Rec      ->  RecF . r . unRecF
>               d @ g    ->  CompF . pmap (fmap2 p r) . unCompF
>               Const t  ->  ConstF . unConstF

%align
%}
%{
%align 12

> cata     :: Regular d => (FunctorOf d a b -> b) -> (d a -> b)
> cata i   =  i   . fmap2 id (cata i  ) . out

> ana      :: Regular d => (b -> FunctorOf d a b) -> (b -> d a)
> ana    o =  inn . fmap2 id (ana    o) . o

> hylo     :: Bifunctor f => (f a b -> b) -> (c -> f a c) -> c -> b
> hylo i o =  i   . fmap2 id (hylo i o) . o

> para     :: Regular d => (d a -> FunctorOf d a b -> b) -> d a -> b
> para i x =  i x (fmap2 id (para i) (out x))

%align
%}

