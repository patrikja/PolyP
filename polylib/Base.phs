---------------------------------------------------------------
-- Basic polytypic functions
--   980409 Patrik Jansson
---------------------------------------------------------------
module Base(pmap,fmap,cata,ana,hylo,para,(-*-),(-+-)) where

pmap :: Regular d => (a -> b) -> d a -> d b
pmap f   = inn . fmap f  (pmap f)   . out

polytypic fmap :: (a -> c) -> (b -> d) -> f a b -> f c d
  = \p r -> case f of                             
      	      g + h   ->  (fmap p r) -+- (fmap p r)
      	      g * h   ->  (fmap p r) -*- (fmap p r)
      	      Empty   ->  id                    
      	      Par     ->  p                     
      	      Rec     ->  r                     
      	      d @ g   ->  pmap (fmap p r)       
      	      Const t ->  id                    
      
cata :: Regular d => (FunctorOf d a b -> b) -> (d a -> b)
ana  :: Regular d => (b -> FunctorOf d a b) -> (b -> d a)
hylo :: Bifunctor f => (f a b -> d)-> (c -> f a b) -> c -> d
cata i   = i   . fmap id (cata i  ) . out
ana    o = inn . fmap id (ana    o) . o
hylo i o = i   . fmap id (hylo i o) . o

para :: Regular d => (d a -> FunctorOf d a b -> b) -> d a -> b
para i x = i x (fmap id (para i) (out x))

---------------------------------------------------------------
-- non-polytypic help functions
(-*-) :: (a -> c) -> (b -> d) -> (a,b) -> (c,d)
(-+-) :: (a -> c) -> (b -> d) -> Either a b -> Either c d

(f -*- g) = \p -> (f (fst p), g (snd p))
(f -+- g) = either (Left . f) (Right . g)
