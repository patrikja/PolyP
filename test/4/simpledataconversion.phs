import Base(fmap2)
-- PJ 990705

dc :: (Regular d, Regular e) => 
      (FunctorOf e a (d b) -> FunctorOf d b (d b)) ->
      e a -> d b
dc f = inn . f . fmap2 id (dc f) . out

flist2ftree :: FunctorOf ([]) a (Tree a) -> FunctorOf Tree a (Tree a)
flist2ftree x = case x of 
  (InL EmptyF)   -> InL EmptyF
  (InR (ParF x :*: RecF y)) -> InR (InR (RecF (Leaf x) :*: RecF y))

data Tree a = Empty | Leaf a | Bin (Tree a) (Tree a) deriving Show

list2tree = dc flist2ftree

main = print (list2tree "Hej")

