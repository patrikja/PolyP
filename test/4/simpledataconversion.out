Bin (Leaf 'H') (Bin (Leaf 'e') (Bin (Leaf 'j') Empty))
